lib/services/bootstrap/safeBootstrap.js

```diff
@@ -0,0 +1,194 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2018 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+/**
+ * Base class for index bootstraping.
+ *
+ * Allow to have a cluster safe bootstrap sequence played on only 1 node and
+ * having other node to wait the end of the sequence.
+ *
+ * Bootstrap sequence have a configurable timeout.
+ */
+class SafeBootstrap {
+  /**
+   * @param {string} name
+   * @param {Kuzzle} kuzzle
+   * @param {InternalEngine} engine
+   * @param {number} timeout
+   */
+  constructor (name, kuzzle, engine, timeout) {
+    this.kuzzle = kuzzle;
+    this.engine = engine;
+    this.attemptDelay = Math.round(timeout / 10);
+
+    // IDs for config documents
+    this._LOCK_ID = `${name}.lock`;
+    this._BOOTSTRAP_DONE_ID = `${name}.done`;
+  }
+
+  /**
+   * Sequence of actions to play for bootstrap
+   * @abstract
+   * @param  {...any} args - Arguments given to startOrWait()
+   * @returns {Promise}
+   */
+  _bootstrapSequence (...args) {
+    throw new Error(`To be implemented. args: ${args.join()}`);
+  }
+
+  /**
+   * Throw an error when the bootstrap timeout exceed
+   * @abstract
+   * @returns {Promise}
+   */
+  _throwLockWaitTimeout () {
+    throw new Error('To be implemented.');
+  }
+
+  /**
+   * Try to start the bootstrap sequence.
+   * If the bootstrap sequence has already been played, resolves the promise.
+   *
+   * Otherwise,
+   *  if the lock is free, the bootstrap sequence will be played from this node
+   *  otherwise, this node will wait until the bootstrap sequence is over.
+   *
+   * This method will reject if the bootstrap exceeds bootstrapLockTimeout value.
+   *
+   * @param  {...any} args - Arguments passed to _bootstrapSequence() method
+   *
+   * @returns {Promise}
+   */
+  async startOrWait (...args) {
+    if (await this.engine.exists('config', this._BOOTSTRAP_DONE_ID)) {
+      return;
+    }
+
+    if (await this._isLocked()) {
+      return this._waitTillUnlocked();
+    }
+
+    return new Promise((resolve, reject) => {
+      this._waitTillUnlocked()
```

Comment:
missing await / chained call to this._playBootstrap

Reply:
It's intentional. If this promise reject, I reject the promise returned by `startOrWait`.  
It's like the pipe's timeout, I want to reject `startOrWait` if `_playBootstrap` takes too much time.

Reply:
Then you should make sure that the promise cannot be rejected twice, which will be the case if _playBootstrap rejects after _waitTillUnlocked.